src/app/navigation/utils.ts

```diff
@@ -109,3 +114,26 @@ export const findPathForParams = (params: object, paths: string[]) => {
       paramsInPath.every(({name}) => paramKeys.includes(name.toString()));
   });
 };
+
+export const getScrollTargetFromQuery = (
+  query: OutputParams,
+  hash: string
+): HighlightScrollTarget | SearchScrollTarget | null => {
+  if (!hash || !query.target || Array.isArray(query.target)) { return null; }
+  try {
+    const parsed = JSON.parse(decodeURIComponent(query.target));
+    if (
+      parsed instanceof Object
+      && (hasSearchScrollTargetParams(parsed) || hasHighlightScrollTargetParams(parsed))
```

Comment:
Is the only difference between a scroll target and scroll target params the presence of `elementId` that is coming from the hash? Maybe it would be enough to just merge the params with hash as soon as possible and validate it as a scroll target?